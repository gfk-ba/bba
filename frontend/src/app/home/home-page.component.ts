import { Store } from '@ngrx/store';
import { StoreComponent } from '../components/base/store-component.class';
import { State } from '../reducers';
import { Greet, GreetWorld } from '../reducers/home/home.actions';
import { greetingSelector } from '../reducers/home/home.selectors';
import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  OnInit,
  OnDestroy,
} from '@angular/core';

@Component({
  selector: 'dcs-home-page',
  templateUrl: './home-page.component.html',
  styleUrls: ['./home-page.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class HomePageComponent extends StoreComponent implements OnInit, OnDestroy {
  public greeting = '';

  constructor(protected store: Store<State>, protected cd: ChangeDetectorRef) {
    super(store, cd);
  }

  /**
   * Will run just before the component will be rendered;
   * best place to load data, subscribe to changes (...)
   */
  public ngOnInit() {
    this.subscribeToState(greetingSelector, greeting => {
      this.greeting = greeting;
    });

    setTimeout(() => {
      this.dispatch(new GreetWorld());
    }, 1000);

    setTimeout(() => {
      this.dispatch(new Greet('DCS'));
    }, 5000);
  }
}
