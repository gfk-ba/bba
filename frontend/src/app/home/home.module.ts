import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { HomePageComponent } from './home-page.component';
import { HomeRoutingModule } from './home-routing.module';

@NgModule({
  imports: [CommonModule, HomeRoutingModule],
  declarations: [HomePageComponent],
})
export class HomeModule {}
