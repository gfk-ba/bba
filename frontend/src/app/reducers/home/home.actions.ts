import { Action } from '@ngrx/store';

export enum HomeActionTypes {
  Greet = '[Home] Greet',
  GreetWorld = '[Home] Greet World',
  GreetUniverse = '[Home] Greet Universe',
}

export class Greet implements Action {
  readonly type = HomeActionTypes.Greet;

  constructor(public payload: string) {}
}

export class GreetWorld implements Action {
  readonly type = HomeActionTypes.GreetWorld;
}

export class GreetUniverse implements Action {
  readonly type = HomeActionTypes.GreetUniverse;
}

export type HomeActions = Greet | GreetWorld | GreetUniverse;
