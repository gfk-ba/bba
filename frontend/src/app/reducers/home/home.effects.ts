import { Injectable } from '@angular/core';
import { Actions, Effect } from '@ngrx/effects';
import { ofType } from '@ngrx/effects';
import { delay, mapTo } from 'rxjs/operators';
import { GreetUniverse, HomeActionTypes } from './home.actions';

@Injectable()
export class HomeEffects {
  @Effect()
  effect$ = this.actions$.pipe(
    ofType(HomeActionTypes.GreetWorld),
    delay(2000),
    mapTo(new GreetUniverse())
  );

  constructor(private actions$: Actions) {}
}
