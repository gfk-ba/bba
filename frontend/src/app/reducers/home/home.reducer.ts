import { HomeActions, HomeActionTypes } from './home.actions';

export interface State {
  greeting: string;
}

export const initialState: State = {
  greeting: 'Unknown',
};

export function reducer(state = initialState, action: HomeActions): State {
  switch (action.type) {
    case HomeActionTypes.Greet:
      return { ...state, greeting: action.payload };

    case HomeActionTypes.GreetWorld:
      return { ...state, greeting: 'World' };

    case HomeActionTypes.GreetUniverse:
      return { ...state, greeting: 'Universe' };

    default:
      return state;
  }
}
