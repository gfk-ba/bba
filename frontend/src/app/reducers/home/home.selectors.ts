import { createFeatureSelector, createSelector } from '@ngrx/store';
import { State } from './home.reducer';

export const homeStateSelector = createFeatureSelector<State>('home');

export const greetingSelector = createSelector([homeStateSelector], state => state.greeting);
