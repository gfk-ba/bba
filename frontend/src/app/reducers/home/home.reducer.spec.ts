import { Greet, GreetUniverse, GreetWorld } from './home.actions';
import { initialState, reducer } from './home.reducer';

describe('Home Reducer', () => {
  describe('unknown action', () => {
    it('should return the initial state', () => {
      const action = {} as any;
      const result = reducer(initialState, action);
      expect(result).toBe(initialState);
    });
  });

  describe('Greet', () => {
    it('puts the payload into greeting', () => {
      const result = reducer(initialState, new Greet('Me'));
      expect(result).toMatchSnapshot();
    });
  });

  describe('GreetWorld', () => {
    it('puts World into greeting', () => {
      const result = reducer(initialState, new GreetWorld());
      expect(result).toMatchSnapshot();
    });
  });

  describe('GreetUniverse', () => {
    it('puts Universe into greeting', () => {
      const result = reducer(initialState, new GreetUniverse());
      expect(result).toMatchSnapshot();
    });
  });
});
