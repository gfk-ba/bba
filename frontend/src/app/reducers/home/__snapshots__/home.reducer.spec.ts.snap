// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Reducer Greet puts the payload into greeting 1`] = `
Object {
  "greeting": "Me",
}
`;

exports[`Home Reducer GreetUniverse puts Universe into greeting 1`] = `
Object {
  "greeting": "Universe",
}
`;

exports[`Home Reducer GreetWorld puts World into greeting 1`] = `
Object {
  "greeting": "World",
}
`;
